Fix Start-Process -Wait polling

jborean93 opened the pull request:
# PR Summary
Update the logic for waiting for a process and its dependencies to use a completion port rather than a spin lock that polls every second. This makes it more efficient to wait for it to complete as it will return straight away.

Also adds stopping support for Unix and `-Wait` and when a job isn't used to track processes as currently they do not respond to `StopProcessing`.

## PR Context
Fixes: https://github.com/PowerShell/PowerShell/issues/24709

The new approach is the one recommended by Raymond Chen at https://devblogs.microsoft.com/oldnewthing/20130405-00/?p=4743. It sets up an IO completion port and monitors the completion codes returned by the job until we've received `JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO` signalling all the processes have ended inside the job.

There should not be any breaking changes here, just an improvement in how we poll for the job completion.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
  - [x] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
  - [x] N/A or can only be tested interactively - too many race conditions to test this fully
  - **OR**
  - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed: <!-- Number/link of that issue here -->

jborean93 committed:
Fix Start-Process -Wait polling

Update the logic for waiting for a process and its dependencies to use a
completion port rather than a spin lock that polls every second. This
makes it more efficient to wait for it to complete as it will return
straight away.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2742,74 +2740,145 @@ private static IEnumerable<string> CompleteFileVerbs(string filePath, string wor
+
+            // // Add the process to the job object
+            return Interop.Windows.AssignProcessToJobObject(
+                this._jobObjectHandle,
```
We can remove the `this`-s for all new private fields and make the code more readable.

jborean93 replied:
I had that originally but one of the CI checks was failing without using this for private fields.

iSazonov replied:
I see https://dev.azure.com/powershell/PowerShell/_build/results?buildId=153772&view=logs&j=69ceb723-764d-5bc2-ea34-985f6795a822&t=738cb80a-7c64-5f29-cdf2-975104f8f46f
> D:\a\1\s\src\Microsoft.PowerShell.Commands.Management\commands\management\Process.cs(2821,36): error CS0029: Cannot implicitly convert type 'bool' to 'nint' [D:\a\1\s\src\Microsoft.PowerShell.Commands.Management\Microsoft.PowerShell.Commands.Management.csproj]

It is not related `this`. I guess there was a typo.

jborean93 replied:
It was the CodeFactor checks and not the actual build. You can actually see one of them in https://www.codefactor.io/repository/github/powershell/powershell/pull/24711 but the fixes outweigh the added "problems" picked up.

>  Prefix local calls with this. 

I can't say I agree but the check is there so I'll just conform to it.

iSazonov replied:
In our Settings.StyleCop file the rule is disabled. I don't know why CodeFactor ignores our config.

And we never follow the rule. So please remove the `this`-s.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2742,74 +2740,145 @@ private static IEnumerable<string> CompleteFileVerbs(string filePath, string wor
+                    out completionCode,
+                    out _,
+                    out _,
+                    INFINITE);
```
> INFINITE

Will does Ctrl-C still kill the cmdlet/pipeline?

jborean93 replied:
That’s a good point, I never hooked that up.

MatejKafka replied:
This SO post contains a reasonable-looking summary of possible ways to wake up the thread: https://stackoverflow.com/questions/32921513/how-to-force-getqueuedcompletionstatus-to-return-immediately

MatejKafka replied:
Given that we only have a single thread on the completion port, I think the cleanest option is to use `PostQueuedCompletionStatus` and post a custom completion to the IOCP.

jborean93 replied:
I've added code to post the same `JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO` completion code when a cancellation is done. This means we don't step on any other completion codes added in the future and a simple is cancelled check means we can differentiate between no processes and cancelled.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2742,74 +2740,145 @@ private static IEnumerable<string> CompleteFileVerbs(string filePath, string wor
+                CompletionPort = this._completionPortHandle,
+            };
+
+            unsafe
+            {
+                this._initStatus = Interop.Windows.SetInformationJobObject(
+                    this._jobObjectHandle,
+                    Interop.Windows.JobObjectAssociateCompletionPortInformation,
+                    (nint)(&completionPort),
+                    Marshal.SizeOf<Interop.Windows.JOBOBJECT_ASSOCIATE_COMPLETION_PORT>());
+            }
```
Please move unsafe code to Interop class. We can create new helper method there and call it here.

jborean93 replied:
Any common location to put these wrapper functions?

iSazonov replied:
Near p/invoke definition. One example FindFirstFile.cs or FindExecutable.cs

jborean93 replied:
Done.

iSazonov reviewed (commented):
Only style comments.

Also please add more info in the PR description: motivation, links to blog posts.

MatejKafka commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2742,74 +2740,145 @@ private static IEnumerable<string> CompleteFileVerbs(string filePath, string wor
-            int dwSize = 0;
-            const int JOB_OBJECT_BASIC_PROCESS_ID_LIST = 3;
-            JOBOBJECT_BASIC_PROCESS_ID_LIST JobList = new();
+            if (this._completionPortHandle == nint.Zero)
```
Any specific reason not to just use `== 0`?

jborean93 replied:
Just personal preference, I prefer to be explicit that we are working with native int types with these checks.

MatejKafka commented on src/System.Management.Automation/engine/Interop/Windows/AssignProcessToJobObject.cs:
```diff
@@ -9,7 +9,7 @@ internal static partial class Interop
     internal static partial class Windows
     {
-        [LibraryImport("Kernel32.dll", SetLastError = true)]
+        [LibraryImport("api-ms-win-core-job-l2-1-0.dll", SetLastError = true)]
```
Are you sure we should depend on the specific API set DLL? Documentation recommends to link with kernel32.dll.

jborean93 replied:
Good point, I'll update them to use only what is documented.

jborean93 committed:
Tidy up code from review, remove some old code no longer needed

jborean93 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2742,74 +2740,145 @@ private static IEnumerable<string> CompleteFileVerbs(string filePath, string wor
     /// child processes created by the main process hosted by the Start-process cmdlet.
     /// </summary>
-    internal class ProcessCollection
+    internal class ProcessCollection : IDisposable
```
Done

jborean93 commented:
@iSazonov someone should look at why code factor is not respecting the rules set by PowerShell. You can see the failures from missing `this.` at https://www.codefactor.io/repository/github/powershell/powershell/pull/24711. There are other failures but now I have no idea if they are ok to ignore or whether they are actual changes to conform to the pwsh repo. I'm not going to spend time fixing things that turn out aren't rules pwsh cares about.

iSazonov commented:
@jborean93 I filtered out two noise issues on CodeFactor site. It seems now they takes into account only .editorconfig file. At least their documentation points only to this file. I will open a new issue here so that someone migrates the Settings.StyleCop rules to this file.

Also Code Factor check is _optional_. Usually maintainers ask contributors to fix some CodeFactor issues. All _mandatory_ rules are in .globalconfig file and violations of the rules will break the build forcing contributor to fix the code.

iSazonov commented on src/System.Management.Automation/engine/Interop/Windows/CreateJobObject.cs:
```diff
@@ -0,0 +1,17 @@
+    internal static partial class Windows
+    {
+        [LibraryImport("kernel32.dll", EntryPoint = "CreateJobObjectW", SetLastError = true)]
+        internal static partial nint CreateJobObject(
```
The same.

jborean93 replied:
Done

iSazonov commented on src/System.Management.Automation/engine/Interop/Windows/SetInformationJobObject.cs:
```diff
@@ -0,0 +1,47 @@
+            nint jobHandle,
+            ref JOBOBJECT_ASSOCIATE_COMPLETION_PORT completionPort)
+        {
+            unsafe
```
See example FindFirstFileW in our interop code how avoid unsafe code.

jborean93 replied:
I personally find it easier to reason with a small unsafe block then having multiple definition for a PInvoke call just to expose the various `ref` overloads. In this case having this overload with the unsafe block means that

+ the caller doesn't have to worry about also passing in the `JobObjectInformationClass` and `cbJobObjectInformationLength`
+ we don't need to duplicate the `LibraryImport` definition

While the first point isn't inherent to this using unsafe it does mean we can avoid the latter through this more specific overload.

iSazonov replied:
LibraryImport automatically involves a source generator to create specific marshaling code. If we allow the SG to process SafeHandle why do we manually process the struct? The SG creates well-tested code for us. Main benefit of the SG is that we automatically get all innovations from .Net developers. With a manual code, we lose that. The point is that using the SG is preferable, as we have already done in other places.

The struct is only for the pinvoke but it is created in main code. Since we already have the helper method I'd expect the struct creation in the method.

The pinvoke method is only used in the helper method so it should be private.

jborean93 replied:
Fair enough, I'll just add that one definition considering we only use it for that right now anyway.

jborean93 committed:
Apply suggestions from code review

Co-authored-by: Ilya <darpa@yandex.ru>

jborean93 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,161 @@
+    /// JobObjectHandle is a reference to the job object used to track
+    /// the child processes created by the main process hosted by the Start-Process cmdlet.
+    /// </summary>
+    private nint _jobObjectHandle;
+
+    /// <summary>
+    /// The completion port handle that is used to monitor job events.
+    /// </summary>
+    private nint _completionPortHandle;
```
I've somewhat moved away from this pattern personally as it can make the interop code a bit annoying to deal with in some circumstances but I'll continue to use the SafeHandles here as suggested.

jborean93 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2062,7 +2062,7 @@ protected override void BeginProcessing()
 
 #if !UNIX
-            ProcessCollection jobObject = null;
+            using JobProcessCollection jobObject = new();
```
It originally way and I can certainly move it back but always initialising it makes the `using` disposal a lot simpler to deal with. As the job and completion handles are created on demand there shouldn't be too much overhead to always creating it and thus I preferred the easier to understand (IMO) code.

jborean93 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -1604,7 +1604,7 @@ private static string MapReturnCodeToErrorMessage(int returnCode)
     public sealed class StartProcessCommand : PSCmdlet, IDisposable
     {
-        private ManualResetEvent _waithandle = null;
+        private readonly CancellationTokenSource _cancellationTokenSource = new();
```
If I made it lazy it is subject to a race condition. If the stop signal was called just before the cancellation token source was created then it will never be cancelled and the cmdlet will continue to run as if it was never stopped. By always having it there we can ensure that if `StopProcessing` was called then it will be marked as cancelled. It's also cleaner code wise as we no longer need the null checks and don't have to reason when the cancellation token is present.

jborean93 commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:
```diff
@@ -2151,23 +2150,20 @@ protected override void BeginProcessing()
                             // WinBlue: 27537 Start-Process -Wait doesn't work in a remote session on Windows 7 or lower.
                             // A Remote session is in it's own job and nested job support was only added in Windows 8/Server 2012.
-                            process.WaitForExit();
+                            process.WaitForExitAsync(_cancellationTokenSource.Token).GetAwaiter().GetResult();
```
Potentially, while it doesn't say so the else branch here also handles other cases where we either failed to create the job object or add a process to the job. I'm unsure if these scenarios could actually happen but we've been bitten before in this code so it's nice to have the fallback. Certainly the comments could be updated though to better reflect what this is now handling.

jborean93 commented:
> Also Code Factor check is optional. Usually maintainers ask contributors to fix some CodeFactor issues. All mandatory rules are in .globalconfig file and violations of the rules will break the build forcing contributor to fix the code.

Thanks for looking into that. If this is optional then really the check should be removed. While this isn't probably the best place to talk about it having what looks to be a failing CI check can be annoying for contributors trying to get their PR in a ready for review state. An optional check is ultimately not a check at all and is extra noise and work for everyone involved. It should either be enforced or just removed altogether.

As for the other comments, I'll push a commit too add the safe handles back in. I've replied to some of the other review comments you've made though as to why I think they shouldn't change. Appreciate you looking through this.

jborean93 committed:
Use SafeHandles instead of raw IntPtr values for the job and completion port

jborean93 commented on src/System.Management.Automation/engine/Interop/Windows/CreateIoCompletionPort.cs:
```diff
@@ -0,0 +1,19 @@
+    internal static unsafe partial class Windows
+    {
+        [LibraryImport("kernel32.dll", SetLastError = true)]
+        internal static partial nint CreateIoCompletionPort(
```
Done

iSazonov commented:
@jborean93 Before we continue the code review in details I want to share some conceptual thoughts.

>  If this is optional then really the check should be removed.

We can not remove CodeFactor because it does useful work for us. For example, look issues for XML documentation comments. Historically pwsh has thousands of empty XML doc comments. If we make CodeFactor mandatory we block whole project since it is not really fix all such issues. We can not disable such issues because we want contributors add XML doc comments in new code. So CodeFactor as "optional" check is a _compromise_. Our workflow is quite gentle - if the maintainer sees significant CodeFactor notices, it will ask contributor to fix them to follow the repository rules.
Also issues the CodeFactor reports comes from StyleCop analyzer. The analyzer works _locally_ for you too. Main problem is CodeFactor has false positives from time to time. We can't influence it. So this is another argument to make it optional.
I hope this information will help make your work here a little more comfortable.

----

Looking on a piece of the method we change here, it is conceptually as simple as:
- create Process object
- write it in pipeline if needed
- wait it if needed

But the real code, due to the previously added workarounds, has become more confusing, and with the current fixes, the situation has become even worse.
We have already fixed this code more than once, and we can assume that this will continue.
In this case, we would like to bring the code to a state more convenient for maintenance.
Now, in addition to the problems already mentioned, there are other amazing things. 
For example, a JobCollection.  Why is this a collection? Obviously, it was designed for use in pipeline, but we only use it only in BeginProcessing for years. It is worth not only renaming this type but also simplifying it.

Another amazing thing is the comment: https://github.com/PowerShell/PowerShell/blob/c0d7fb787f025dffdcc4ea8a05e65004e9e71d66/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L2093-L2100
For some reason, this comment is located after the code it refers to.

Next amazing thing is: https://github.com/PowerShell/PowerShell/blob/c0d7fb787f025dffdcc4ea8a05e65004e9e71d66/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L2133-L2150
The `process` is created by constructor (throw on error) and cannot be null. So no need to check this and write terminating error (after `if (Wait.IsPresent)` we do the same too).

Perhaps I didn't see all the problems when I skimmed through this method.

And I wouldn't write all this if you had already fixed this code more than once and probably would again. So I would welcome converting this code into something cleaner and more maintainable.
At first glance, it would be possible to create a `ProcessWating` wrapper for `Process` with different implementations for Windows and Unix. This would make the main code very simple. The wrapper could be just a disposable struct.

jborean93 commented:
> and with the current fixes, the situation has become even worse.

While yes the current code is a bit of a mess I'm not sure I agree that this makes it worse. This change

+ Moves out a nested class
+ Moves out some PInvoke definitions, even some that was defined in another area for some reason
+ Simplifies some nullability checks and object initialisation
+ Removes some null condition checks around the wait handle to the more common cancellation token setup

Yes some work should be done to tidy the code up and probably unpick the attempts to share some of the code flow into their own specific paths but I don't think it should be done in the same PR. I'm happy to open another PR to try and clean some more stuff up later on though.

> We have already fixed this code more than once, and we can assume that this will continue.

We've fixed similar code but not this issue, this specific issue around jobs polling every 1 second has always been around.

> For some reason, this comment is located after the code it refers to.

The comment is directly for the code just below it

```csharp
if (Wait)
{
    jobObject = new();
    jobAssigned = jobObject.AssignProcessToJobObject(processInfo.Process);
}
```

It is referring to the fact that `AssignProcessToJobObject` might fail unless we have the `ALL_ACCESS` token from creating the process. In the past we did

```csharp
Process process = Process.GetProcessById(processInfo.ProcessId);
jobObject.AssignProcessToJobObject(process.SafeHandle);
```

This was problematic because the handle returned by `process.SafeHandle` may not have enough permissions in the case when we started the process with custom credentials.

> The process is created by constructor (throw on error) and cannot be null. So no need to check this and write terminating error (after if (Wait.IsPresent) we do the same too).

I've not looked at this code and think it's unrelated to the problem at hand.

> And I wouldn't write all this if you had already fixed this code more than once and probably would again. So I would welcome converting this code into something cleaner and more maintainable.

I certainly agree that this should be fixed but I don't think it should be rewritten at the same time as fixing an unrelated problem. Refactoring should only happen when you want to cleanup the code, not do bugfixes at the same time.

iSazonov commented:
/azp run

jborean93 committed:
Use specific struct overload for SetInformationJobObject

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,143 @@
+
+        if (_jobObject is null)
+        {
+            _jobObject = Interop.Windows.CreateJobObject(nint.Zero, nint.Zero);
```
Usually we use named arguments for constants. But the zero arguments don't add any value here. There are examples in our interop code where we create helper to hide extra parameters. I suggest to follow the pattern.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,143 @@
+                lpCompletionKey: out _,
+                lpOverlapped: out _,
+                INFINITE);
+        }
```
The same. Here we only use first two parameters. We should hide the rest.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,143 @@
+                Interop.Windows.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO,
+                lpCompletionKey: nint.Zero,
+                lpOverlapped: nint.Zero);
+        });
```
The same.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,143 @@
+                FileHandle: -1,
+                ExistingCompletionPort: nint.Zero,
+                CompletionKey: nint.Zero,
+                NumberOfConcurrentThreads: 1);
```
The same.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,143 @@
+    ~JobProcessCollection()
+    {
+        Dispose();
+    }
```
Extra method. We can remove since all unmanaged resources wrapped in SafeHandle.
https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern#finalizable-types

jborean93 committed:
Changes from review

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,123 @@
+
+        if (_jobObject is null)
+        {
+            _jobObject = Interop.Windows.CreateJobObject();
+            if (_jobObject.IsInvalid)
+            {
+                _initStatus = false;
+                return false;
```
```suggestion
            _jobObject = Interop.Windows.CreateJobObject();
            if (_jobObject.IsInvalid)
            {
                _initStatus = false;
                jobObject.Dispose();
                jobObject = null;
                return false;
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,123 @@
+
+        if (_completionPort is null)
+        {
+            _completionPort = Interop.Windows.CreateIoCompletionPort();
+
+            if (_completionPort.IsInvalid)
+            {
+                _initStatus = false;
+                return false;
+            }
```
We could call `WaitForExit()` with invalid handle.
```suggestion
            _completionPort = Interop.Windows.CreateIoCompletionPort();
            if (_completionPort.IsInvalid)
            {
                _initStatus = false;
                _completionPort.Dispose();
                _completionPort = null;
                return false;
            }
```

iSazonov commented on src/System.Management.Automation/engine/Interop/Windows/CreateIoCompletionPort.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        internal sealed class SafeIoCompletionPort : SafeHandle
+        {
+            public SafeIoCompletionPort() : base(nint.Zero, true) { }
```
```suggestion
            public SafeIoCompletionPort() : base(invalidHandleValue: nint.Zero, ownsHandle: true) { }
```

jborean93 replied:
Seems a bit overkill when we don't seem to do it anywhere else but ok.

iSazonov replied:
Just style for readability.

iSazonov commented on src/System.Management.Automation/engine/Interop/Windows/CreateJobObject.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        internal sealed class SafeJobHandle : SafeHandle
+        {
+            public SafeJobHandle() : base(nint.Zero, true) { }
```
```suggestion
            public SafeJobHandle() : base(invalidHandleValue: nint.Zero, ownsHandle: true) { }
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/JobProcessCollection.cs:
```diff
@@ -0,0 +1,123 @@
+        {
+            Interop.Windows.GetQueuedCompletionStatus(
+                _completionPort,
+                INFINITE,
```
Please remove the const and move it to the helper.

jborean93 replied:
Done

iSazonov reviewed (commented):
@jborean93 LGTM with same style comments.
Please don't rebase without maintainer ask. Rebase kill commit history and resets previous reviews.

jborean93 commented:
> Please don't rebase without maintainer ask. Rebase kill commit history and resets previous reviews.

CI says that the branch is out of date and I rebase to bring it back up to date. Using a merge pull creates merge commits and while yes a rebase requires a force push it still preserves the commit history for the PR. If you have a better suggestion then please share, personally I'm getting sick of PowerShell's CI that marks PRs as red or not ready when some things seem to be optional.

I still see the reviews, I marked them as resolved because I responded to the latest comments on there, sorry if I missed anything.

jborean93 reviewed (commented):
Edit: oops I thought this was the commit window.

jborean93 committed:
Apply suggestions from code review

Co-authored-by: Ilya <darpa@yandex.ru>

jborean93 committed:
Move const to interop and fix typo

iSazonov commented:
@jborean93 MSFT team use a tool for reviewing _commit by commit_ and they ask us to keep commit history. GitHub Web GUI also doesn't like rebase - it can hide some comments (and it is very tricky to find them).
Of course, there are situations (very rarely) when we are forced to do a rebase, but at the same time we must protect the time spent by others. So it's better to coordinate rebase.

"Update branch with merge commit" button is not useful for us since we always do "Rebase and Merge" finally. I have not permission to remove the option. "Update branch with rebase" could be useful for us (mainly maintainers) if we need to rebase for running CIs before merge of very old PR.

iSazonov commented:
/azp run

jborean93 commented:
Thanks for the review and merge.

iSazonov commented:
@jborean93 Thanks for great work! I hope you will find the time to clean up the file(s) so that it will be easier for us to do the subsequent work.
